File: Stack and Queue/20_Valid-Parentheses/20_Valid-Parentheses.py
# !/usr/bin/env python 
# -*- coding:utf-8 -*-
"""
Created on Fri Nov  1 16:38:58 2019

@author: SHI YI
"""

class Solution:
    def isValid(self, s: str) -> bool:
        left = "({["
        right = ")}]"
        stack = []  # empty list, used as stack

        for c in s:
            if c in left:     # add c to ls, if c in 'left'
                stack.append(c)
            elif c in right:  # skip c, if c in 'right' 
                if len(stack) == 0:
                    return False
                elif right.index(c) != left.index(stack.pop()):
                    return False
        return len(stack) == 0


s = "("

solve = Solution()

if solve.isValid(s):
    print("True")
else:
    print("False")
    

File: Stack and Queue/20_Valid-Parentheses/20_Valid-Parentheses.cpp
/*
Title:
    20. Valid Parentheses
    20. 有效括号
Address:
    https://leetcode-cn.com/problems/valid-parentheses/
*/


#include <iostream>
#include <vector>
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s){
        stack<char> stk;

        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == ')' || s[i] == ']' || s[i] == '}') {
                if (stk.empty() || !isLegal(stk.top(), s[i]))
                    return false;
                stk.pop();
            }
            else
                stk.push(s[i]);
        }
        if (!stk.empty())
            return false;
        return true;
    }

    bool isLegal(const char& a, const char& b){
        if (a == '(' && b == ')')
            return true;
        if (a == '[' && b == ']')
            return true;
        if (a == '{' && b == '}')
            return true;
		return false;
    }
};



int main()
{
	// string str {"([}{])"};
	string str {"()"};
	Solution solve;

	cout << solve.isValid(str) << endl;

    return 0;
}
